File: my-folder/3434-find-the-number-of-distinct-colors-among-the-balls/solution.cpp
class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        int n = queries.size();
        unordered_map<int,int> balls;
        unordered_map<int,set<int>> colors;

        vector<int> output;

        for (auto& q:queries)
        {
            int ball = q[0];
            int color = q[1];

            int prev_color = -1;
            if (balls.find(ball)!=balls.end()){
                prev_color = balls[ball];
                if (prev_color == color){
                    output.push_back(colors.size());
                    continue;
                }
            }
            balls[ball] = color;
            colors[color].insert(ball);
            if (prev_color!=-1){
                colors[prev_color].erase(ball);
                if ( colors[prev_color].empty())
                {
                    colors.erase(prev_color);
                } 
            }

            output.push_back(colors.size());

        }
        return output;

    }
};

File: my-folder/3434-find-the-number-of-distinct-colors-among-the-balls/solution.py
class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        n = len(queries)
        balls = defaultdict(int)
        colors = defaultdict(set)

        output = [0]*n

        for i,(ball,color) in enumerate(queries):
            # add color to ball
            prev_color = -1
            if (balls[ball]!=0): # if a ball already has some color
                if (balls[ball] == color):
                    output[i] = len(colors)
                    continue
                prev_color = balls[ball]
            balls[ball] = color 

            # add distinct colors:
            colors[color].add(ball)
            # remove previous color
            if prev_color!=-1:
                colors[prev_color].discard(ball)
                if len (colors[prev_color]) == 0:
                    del colors[prev_color]
            output[i] = len(colors)
        return output
                    
                
                
                
